### DISCUSSION TITLE
using useFormContext to use setValue triggers re-render on other fields

### DISCUSSION THREAD

**Discussion by pooyarm** [react-hook-form/react-hook-form#7440] [Q&A] [answered]
I'm migrating a big and complex form to `react-hook-form` because performance is really important in this case. Imagine that I have a FormProvider on top then, on nested components I've fields (using useFormControl), everything is good, except two of the fields are related to each other, when one of them have an specific value, the value of other one should be updated, I've used `watch` and `setValue` but the problem is that the rest of fields are also re-rendered.

Here is an example: https://codesandbox.io/s/re-rendering-independent-input-f752z?file=/src/App.js

The use case is that when inclusive checkbox is marked price should be reset to zero, which is ok, but there is another third field (IndependentField) which is re-rendered on every change on inclusive checkbox.

of course it can be somehow fixed by using `onChange` but I don't like it because in that case we're spreading the logic of one field into different places.

Does anyone have any idea how this can be fixed smoothly?

**Comment by fahadsohail482** [comment, marked as answer]
@pooyarm you are using the watch(). Consider using the **useWatch** Hook which Behaves similarly to the watch API, however, this will isolate re-rendering at the component level and potentially result in better performance for your application.
Here is one of the Rule regarding watch.

> This API will trigger re-render at the root of your app or form, consider using a callback or the useWatch api if you are experiencing performance issues.

Your Example works fine with useWatch Hook.
Hope this helps.
Here is the Working [Example](https://codesandbox.io/s/re-rendering-independent-input-forked-xjwlv?file=/src/Price.js).

**Reply by pooyarm** [reply]
I've already checked `useWatch`, but it was my mistake because another field was also relying on this field and that was the main reason of re-renderings.

Thanks for the hint.
